File: llm_proxy/records.py
"""Helper functions for parsing LLM request/response records.

The logger uses these functions to determine which type of API a request belongs to, extract request bodies, generate content fingerprints, and find response IDs from responses. They don't write files themselves, only responsible for "understanding record content".
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Mapping

from .payloads import body_json_value


def stable_hash(value: object, length: int = 12) -> str:
    """Generate a stable short hash for any JSON-serializable object.

    ``sort_keys=True`` ensures that dictionaries with different field orders produce the same hash.
    """
    encoded = json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(encoded.encode("utf-8", errors="replace")).hexdigest()[:length]


def safe_filename_part(value: object, fallback: str = "unknown", limit: int = 80) -> str:
    """Convert an arbitrary value into a filename-safe fragment."""
    text = str(value or "").strip()
    safe = "".join(ch if ch.isalnum() else "-" for ch in text).strip("-")
    return (safe[:limit] or fallback).strip("-") or fallback


def get_nested_value(value: object, path: tuple[str, ...]) -> object | None:
    """Safely read a value from a nested dictionary."""
    current = value
    for key in path:
        if not isinstance(current, Mapping):
            return None
        current = current.get(key)
    return current


def first_string(*values: object) -> str | None:
    """Return the first non-empty string."""
    for value in values:
        if isinstance(value, str) and value.strip():
            return value.strip()
    return None


def request_body_json(record: Mapping[str, object]) -> object:
    """Extract the request body from the record and parse it according to JSON/SSE rules."""
    request = record.get("request")
    if not isinstance(request, Mapping):
        return None
    body = request.get("body")
    if not isinstance(body, Mapping):
        return None
    return body_json_value(body)


def response_body_json(record: Mapping[str, object]) -> object:
    """Extract the response body from the record and parse it according to JSON/SSE rules."""
    response = record.get("response")
    if not isinstance(response, Mapping):
        return None
    body = response.get("body")
    if not isinstance(body, Mapping):
        return None
    return body_json_value(body)


def request_path(record: Mapping[str, object]) -> str:
    """Read the client request path from the record."""
    request = record.get("request")
    if not isinstance(request, Mapping):
        return ""
    return str(request.get("path", ""))


def endpoint_kind(path: str) -> str:
    """Determine which type of LLM API based on the path."""
    lowered = path.lower().split("?", 1)[0].rstrip("/")
    if lowered.endswith("/responses") or lowered == "/responses":
        return "responses"
    if lowered.endswith("/messages") or lowered == "/messages":
        return "messages"
    if lowered.endswith("/chat/completions") or lowered == "/chat/completions":
        return "chat"
    if lowered.endswith("/completions") or lowered == "/completions":
        return "completions"
    return "other"


def message_text(value: object) -> object:
    """Format message content into a stable structure for generating fingerprints."""
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        return [message_text(item) for item in value]
    if isinstance(value, Mapping):
        return {key: message_text(value[key]) for key in sorted(value)}
    return value


def _content_text(value: object) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        parts: list[str] = []
        for item in value:
            if isinstance(item, Mapping):
                text = item.get("text")
                if isinstance(text, str):
                    parts.append(text)
            elif isinstance(item, str):
                parts.append(item)
        return "\n".join(parts)
    if isinstance(value, Mapping):
        text = value.get("text")
        if isinstance(text, str):
            return text
    return ""


def is_task_context_message(item: object) -> bool:
    if not isinstance(item, Mapping):
        return False
    text = _content_text(item.get("content")).lstrip()
    fixed_prefixes = (
        "<environment_context>",
        "<permissions instructions>",
        "<app-context>",
        "# Codex desktop context",
    )
    return any(text.startswith(prefix) for prefix in fixed_prefixes)


def chat_system_messages(payload: Mapping[str, object]) -> list[object]:
    """Extract system/developer messages from Chat Completions requests."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return []
    system_roles = {"system", "developer"}
    return [
        {"role": message.get("role"), "content": message_text(message.get("content"))}
        for message in messages
        if isinstance(message, Mapping) and message.get("role") in system_roles
    ]


def chat_prefix_messages(payload: Mapping[str, object], limit: int = 4) -> list[object]:
    """Extract the first few messages from Chat Completions as content fingerprints."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return []
    compacted: list[object] = []
    for message in messages:
        if not isinstance(message, Mapping):
            continue
        if is_task_context_message(message):
            continue
        compacted.append(
            {
                "role": message.get("role"),
                "content": message_text(message.get("content")),
                "name": message.get("name"),
                "tool_call_id": message.get("tool_call_id"),
            }
        )
        if len(compacted) >= limit:
            break
    return compacted


def chat_content_messages(payload: Mapping[str, object]) -> list[object]:
    """Return non-fixed chat messages used as whole-request context evidence."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return []
    compacted: list[object] = []
    for message in messages:
        if not isinstance(message, Mapping):
            continue
        if is_task_context_message(message):
            continue
        compacted.append(
            {
                "role": message.get("role"),
                "content": message_text(message.get("content")),
                "name": message.get("name"),
                "tool_call_id": message.get("tool_call_id"),
            }
        )
    return compacted


def chat_first_user_message(payload: Mapping[str, object]) -> object | None:
    """Extract the first user message from Chat Completions."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return None
    for message in messages:
        if isinstance(message, Mapping) and message.get("role") == "user" and not is_task_context_message(message):
            return message_text(message.get("content"))
    return None


def responses_input_items(payload: Mapping[str, object]) -> list[object]:
    """Convert Responses API input into a uniform list."""
    input_value = payload.get("input")
    if isinstance(input_value, list):
        return input_value
    if input_value is None:
        return []
    return [input_value]


def responses_input_item_summary(item: object) -> object:
    """Compress a single Responses input item, keeping only core fields for task identification."""
    if isinstance(item, str):
        return item
    if not isinstance(item, Mapping):
        return item
    summary: dict[str, object] = {}
    for key in ("type", "role", "call_id", "name"):
        value = item.get(key)
        if value is not None:
            summary[key] = value

    content = item.get("content")
    if isinstance(content, list):
        compact_content = []
        for content_item in content:
            if not isinstance(content_item, Mapping):
                compact_content.append(content_item)
                continue
            compact_entry = {
                "type": content_item.get("type"),
                "text": content_item.get("text"),
                "arguments": content_item.get("arguments"),
                "call_id": content_item.get("call_id"),
            }
            # Remove fields with None values to make fingerprints more stable and concise.
            compact_content.append({key: value for key, value in compact_entry.items() if value is not None})
        summary["content"] = compact_content
    elif content is not None:
        summary["content"] = message_text(content)

    for key in ("output", "arguments"):
        value = item.get(key)
        if value is not None:
            summary[key] = message_text(value)
    return summary or message_text(item)


def responses_input_prefix(payload: Mapping[str, object], limit: int = 6) -> list[object]:
    """Extract the first few input items from Responses API as content fingerprints."""
    content_items = [item for item in responses_input_items(payload) if not is_task_context_message(item)]
    return [responses_input_item_summary(item) for item in content_items[:limit]]


def responses_first_user_message(payload: Mapping[str, object]) -> object | None:
    """Extract the first user text from Responses API input."""
    for item in responses_input_items(payload):
        if not isinstance(item, Mapping) or item.get("role") != "user":
            continue
        if is_task_context_message(item):
            continue
        content = item.get("content")
        if isinstance(content, list):
            texts = []
            for content_item in content:
                if isinstance(content_item, Mapping):
                    text = content_item.get("text")
                    if isinstance(text, str) and text:
                        texts.append(text)
            if texts:
                return texts
        elif content:
            return message_text(content)
    return None


def responses_user_messages(payload: Mapping[str, object]) -> list[object]:
    """Return non-fixed user messages from a Responses API payload."""
    user_messages: list[object] = []
    for item in responses_input_items(payload):
        if not isinstance(item, Mapping) or item.get("role") != "user":
            continue
        if is_task_context_message(item):
            continue
        user_messages.append(responses_input_item_summary(item))
    return user_messages


def chat_user_messages(payload: Mapping[str, object]) -> list[object]:
    """Return non-fixed user messages from a Chat Completions payload."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return []
    user_messages: list[object] = []
    for message in messages:
        if not isinstance(message, Mapping) or message.get("role") != "user":
            continue
        if is_task_context_message(message):
            continue
        user_messages.append(
            {
                "role": message.get("role"),
                "content": message_text(message.get("content")),
                "name": message.get("name"),
            }
        )
    return user_messages


def claude_system_messages(payload: Mapping[str, object]) -> list[object]:
    """Extract top-level system messages from Anthropic/Claude Messages requests."""
    system = payload.get("system")
    if not system:
        return []
    return [{"role": "system", "content": message_text(system)}]


def claude_message_summary(message: Mapping[str, object]) -> dict[str, object]:
    """Compress a Claude message while preserving role and content shape."""
    summary: dict[str, object] = {
        "role": message.get("role"),
        "content": message_text(message.get("content")),
    }
    for key in ("name", "tool_use_id"):
        value = message.get(key)
        if value is not None:
            summary[key] = value
    return summary


def claude_messages(payload: Mapping[str, object]) -> list[object]:
    """Return non-fixed messages from an Anthropic/Claude Messages payload."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return []
    compacted: list[object] = []
    for message in messages:
        if not isinstance(message, Mapping):
            continue
        if is_task_context_message(message):
            continue
        compacted.append(claude_message_summary(message))
    return compacted


def claude_prefix_messages(payload: Mapping[str, object], limit: int = 4) -> list[object]:
    """Extract the first few Claude messages as content fingerprints."""
    return claude_messages(payload)[:limit]


def claude_first_user_message(payload: Mapping[str, object]) -> object | None:
    """Extract the first user message from an Anthropic/Claude Messages request."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return None
    for message in messages:
        if isinstance(message, Mapping) and message.get("role") == "user" and not is_task_context_message(message):
            return message_text(message.get("content"))
    return None


def claude_user_messages(payload: Mapping[str, object]) -> list[object]:
    """Return non-fixed user messages from an Anthropic/Claude Messages payload."""
    messages = payload.get("messages")
    if not isinstance(messages, list):
        return []
    user_messages: list[object] = []
    for message in messages:
        if not isinstance(message, Mapping) or message.get("role") != "user":
            continue
        if is_task_context_message(message):
            continue
        user_messages.append(claude_message_summary(message))
    return user_messages


def request_user_messages(kind: str, payload: object) -> list[object]:
    """Extract the user-message sequence used to decide task continuation."""
    if not isinstance(payload, Mapping):
        return []
    if kind == "responses":
        return responses_user_messages(payload)
    if kind == "chat":
        return chat_user_messages(payload)
    if kind == "messages":
        return claude_user_messages(payload)
    if kind == "completions":
        prompt = payload.get("prompt")
        return [message_text(prompt)] if prompt else []
    return []


def request_fingerprints(kind: str, payload: object) -> dict[str, str]:
    """为不同接口类型生成请求指纹。

    指纹用于判断两个请求是否可能属于同一个任务。不同接口的字段结构不同，
    所以这里分别处理 responses/chat/messages/completions。
    """
    if not isinstance(payload, Mapping):
        return {}

    fingerprints: dict[str, str] = {}
    if kind == "responses":
        instructions = payload.get("instructions")
        if instructions:
            fingerprints["instructions"] = stable_hash(instructions)
        tools = payload.get("tools")
        if tools:
            fingerprints["tools"] = stable_hash(tools)
        first_user = responses_first_user_message(payload)
        if first_user:
            fingerprints["first_user"] = stable_hash(first_user)
        input_prefix = responses_input_prefix(payload)
        if input_prefix:
            fingerprints["input_prefix"] = stable_hash(input_prefix)
        input_value = payload.get("input")
        if input_value:
            fingerprints["input"] = stable_hash(input_value)
    elif kind == "chat":
        system_messages = chat_system_messages(payload)
        if system_messages:
            fingerprints["system"] = stable_hash(system_messages)
        prefix_messages = chat_prefix_messages(payload)
        if prefix_messages:
            fingerprints["messages_prefix"] = stable_hash(prefix_messages)
        content_messages = chat_content_messages(payload)
        if content_messages:
            fingerprints["messages"] = stable_hash(content_messages)
        first_user = chat_first_user_message(payload)
        if first_user:
            fingerprints["first_user"] = stable_hash(first_user)
        tools = payload.get("tools", payload.get("functions"))
        if tools:
            fingerprints["tools"] = stable_hash(tools)
    elif kind == "messages":
        system_messages = claude_system_messages(payload)
        if system_messages:
            fingerprints["system"] = stable_hash(system_messages)
        prefix_messages = claude_prefix_messages(payload)
        if prefix_messages:
            fingerprints["messages_prefix"] = stable_hash(prefix_messages)
        content_messages = claude_messages(payload)
        if content_messages:
            fingerprints["messages"] = stable_hash(content_messages)
        first_user = claude_first_user_message(payload)
        if first_user:
            fingerprints["first_user"] = stable_hash(first_user)
        tools = payload.get("tools")
        if tools:
            fingerprints["tools"] = stable_hash(tools)
    elif kind == "completions":
        prompt = payload.get("prompt")
        if prompt:
            fingerprints["prompt"] = stable_hash(prompt)
    return fingerprints


def request_boundary_fingerprints(kind: str, payload: object) -> dict[str, str]:
    """Fingerprints that must not change within one task."""
    fingerprints = request_fingerprints(kind, payload)
    if kind == "responses":
        boundary_keys = {"instructions", "first_user"}
    elif kind in {"chat", "messages"}:
        boundary_keys = {"system", "first_user"}
    elif kind == "completions":
        boundary_keys = {"prompt"}
    else:
        boundary_keys = set()
    return {key: value for key, value in fingerprints.items() if key in boundary_keys}


def response_ids_from_body(body: object) -> list[str]:
    """从响应体中提取可能的响应 ID，并去重。"""
    ids: list[str] = []
    if isinstance(body, Mapping):
        response_id = body.get("id")
        if isinstance(response_id, str) and response_id:
            ids.append(response_id)
        response_payload = body.get("response")
        if isinstance(response_payload, Mapping):
            nested_id = response_payload.get("id")
            if isinstance(nested_id, str) and nested_id:
                ids.append(nested_id)
    return list(dict.fromkeys(ids))

File: src/proxy/records.ts
import { createHash } from "node:crypto";

import { isRecord, stableJsonStringify } from "../shared/index.js";

export type EndpointKind = "responses" | "messages" | "chat" | "completions" | "other";

export function stableHash(value: unknown, length = 12): string {
  return createHash("sha256")
    .update(stableJsonStringify(value), "utf8")
    .digest("hex")
    .slice(0, length);
}

const CLAUDE_TRANSIENT_TASK_FIELDS = new Set(["cache_control"]);

export function endpointKind(path: string): EndpointKind {
  const lowered = path.toLowerCase().split("?", 1)[0]?.replace(/\/+$/, "") ?? "";
  if (lowered === "/responses" || lowered.endsWith("/responses")) {
    return "responses";
  }
  if (lowered === "/messages" || lowered.endsWith("/messages")) {
    return "messages";
  }
  if (lowered === "/chat/completions" || lowered.endsWith("/chat/completions")) {
    return "chat";
  }
  if (lowered === "/completions" || lowered.endsWith("/completions")) {
    return "completions";
  }
  return "other";
}

export function displayEndpoint(path: string | number | boolean | null | undefined): string {
  const value = (path ? String(path) : "").split("?", 1)[0]?.replace(/\/+$/, "") ?? "";
  return value || "/";
}

export function requestMessageCount(kind: EndpointKind, payload: unknown): number | undefined {
  if (!isRecord(payload)) {
    return undefined;
  }
  if (kind === "responses") {
    const input = payload["input"];
    const count = Array.isArray(input)
      ? input.length
      : input === null || input === undefined
        ? 0
        : 1;
    return count + (isPythonTruthy(payload["instructions"]) ? 1 : 0);
  }
  if (kind === "messages") {
    const system = payload["system"];
    const systemCount = Array.isArray(system) ? system.length : isPythonTruthy(system) ? 1 : 0;
    const messages = payload["messages"];
    return systemCount + (Array.isArray(messages) ? messages.length : 0);
  }
  if (kind === "chat") {
    const messages = payload["messages"];
    return Array.isArray(messages) ? messages.length : 0;
  }
  if (kind === "completions") {
    const prompt = payload["prompt"];
    return Array.isArray(prompt) ? prompt.length : prompt === null || prompt === undefined ? 0 : 1;
  }
  const messages = payload["messages"];
  if (Array.isArray(messages)) {
    return messages.length;
  }
  const input = payload["input"];
  if (Array.isArray(input)) {
    return input.length;
  }
  return input === null || input === undefined ? undefined : 1;
}

export function responseTokenCount(payload: unknown): number | undefined {
  const usage = responseUsage(payload);
  if (!isRecord(usage)) {
    return undefined;
  }
  const total = usage["total_tokens"];
  if (typeof total === "number" && Number.isInteger(total)) {
    return total;
  }
  const counts = responseTokenCounts(payload);
  const values = [counts.request, counts.response].filter(
    (value): value is number => value !== undefined,
  );
  return values.length === 0 ? undefined : values.reduce((sum, value) => sum + value, 0);
}

export interface ResponseTokenCounts {
  readonly request: number | undefined;
  readonly response: number | undefined;
}

export function responseTokenCounts(payload: unknown): ResponseTokenCounts {
  const usage = responseUsage(payload);
  if (!isRecord(usage)) {
    return { request: undefined, response: undefined };
  }
  const input = tokenValue(usage, "input_tokens") ?? tokenValue(usage, "prompt_tokens");
  const cachedInputValues = ["cache_creation_input_tokens", "cache_read_input_tokens"]
    .map((key) => tokenValue(usage, key))
    .filter((value): value is number => value !== undefined);
  const cachedInput = cachedInputValues.reduce((sum, value) => sum + value, 0);
  return {
    request:
      input === undefined && cachedInputValues.length === 0
        ? undefined
        : (input ?? 0) + cachedInput,
    response: tokenValue(usage, "output_tokens") ?? tokenValue(usage, "completion_tokens"),
  };
}

export function responseIdsFromBody(body: unknown): string[] {
  if (!isRecord(body)) {
    return [];
  }
  const ids: string[] = [];
  const responseId = body["id"];
  if (typeof responseId === "string" && responseId !== "") {
    ids.push(responseId);
  }
  const response = body["response"];
  const nestedId = isRecord(response) ? response["id"] : undefined;
  if (typeof nestedId === "string" && nestedId !== "" && !ids.includes(nestedId)) {
    ids.push(nestedId);
  }
  return ids;
}

export function requestFingerprints(kind: EndpointKind, payload: unknown): Record<string, string> {
  if (!isRecord(payload)) {
    return {};
  }
  const fingerprints: Record<string, string> = {};
  if (kind === "responses") {
    for (const key of ["instructions", "tools"] as const) {
      if (isPythonTruthy(payload[key])) {
        fingerprints[key] = stableHash(payload[key]);
      }
    }
    const firstUser = responsesFirstUserMessage(payload);
    if (isPythonTruthy(firstUser)) {
      fingerprints["first_user"] = stableHash(firstUser);
    }
    const inputPrefix = responsesInputPrefix(payload);
    if (inputPrefix.length > 0) {
      fingerprints["input_prefix"] = stableHash(inputPrefix);
    }
    if (isPythonTruthy(payload["input"])) {
      fingerprints["input"] = stableHash(payload["input"]);
    }
  } else if (kind === "chat") {
    const systemMessages = chatSystemMessages(payload);
    if (systemMessages.length > 0) {
      fingerprints["system"] = stableHash(systemMessages);
    }
    const prefixMessages = chatMessages(payload, 4);
    if (prefixMessages.length > 0) {
      fingerprints["messages_prefix"] = stableHash(prefixMessages);
    }
    const contentMessages = chatMessages(payload);
    if (contentMessages.length > 0) {
      fingerprints["messages"] = stableHash(contentMessages);
    }
    const firstUser = chatFirstUserMessage(payload);
    if (isPythonTruthy(firstUser)) {
      fingerprints["first_user"] = stableHash(firstUser);
    }
    const tools = Object.hasOwn(payload, "tools") ? payload["tools"] : payload["functions"];
    if (isPythonTruthy(tools)) {
      fingerprints["tools"] = stableHash(tools);
    }
  } else if (kind === "messages") {
    const systemMessages = claudeSystemMessages(payload);
    if (systemMessages.length > 0) {
      fingerprints["system"] = stableHash(systemMessages);
    }
    const prefixMessages = claudeMessages(payload).slice(0, 4);
    if (prefixMessages.length > 0) {
      fingerprints["messages_prefix"] = stableHash(prefixMessages);
    }
    const contentMessages = claudeMessages(payload);
    if (contentMessages.length > 0) {
      fingerprints["messages"] = stableHash(contentMessages);
    }
    const firstUser = claudeFirstUserMessage(payload);
    if (isPythonTruthy(firstUser)) {
      fingerprints["first_user"] = stableHash(firstUser);
    }
    if (isPythonTruthy(payload["tools"])) {
      fingerprints["tools"] = stableHash(claudeTaskValue(payload["tools"]));
    }
  } else if (kind === "completions" && isPythonTruthy(payload["prompt"])) {
    fingerprints["prompt"] = stableHash(payload["prompt"]);
  }
  return fingerprints;
}

export function requestBoundaryFingerprints(
  kind: EndpointKind,
  payload: unknown,
): Record<string, string> {
  const fingerprints = requestFingerprints(kind, payload);
  const boundaryKeys =
    kind === "responses"
      ? new Set(["instructions", "first_user"])
      : kind === "chat" || kind === "messages"
        ? new Set(["system", "first_user"])
        : kind === "completions"
          ? new Set(["prompt"])
          : new Set<string>();
  return Object.fromEntries(Object.entries(fingerprints).filter(([key]) => boundaryKeys.has(key)));
}

export function requestUserMessages(kind: EndpointKind, payload: unknown): unknown[] {
  if (!isRecord(payload)) {
    return [];
  }
  if (kind === "responses") {
    return responsesInputItems(payload)
      .filter((item) => isRecord(item) && item["role"] === "user" && !isTaskContextMessage(item))
      .map((item) => responsesInputItemSummary(item));
  }
  if (kind === "chat") {
    const messages = payload["messages"];
    const userMessages: unknown[] = [];
    for (const message of Array.isArray(messages) ? (messages as unknown[]) : []) {
      if (isRecord(message) && message["role"] === "user" && !isTaskContextMessage(message)) {
        userMessages.push({
          role: message["role"] ?? null,
          content: messageText(message["content"]),
          name: message["name"] ?? null,
        });
      }
    }
    return userMessages;
  }
  if (kind === "messages") {
    const messages = payload["messages"];
    const userMessages: unknown[] = [];
    for (const message of Array.isArray(messages) ? (messages as unknown[]) : []) {
      if (isRecord(message) && message["role"] === "user" && !isTaskContextMessage(message)) {
        userMessages.push(claudeMessageSummary(message));
      }
    }
    return userMessages;
  }
  if (kind === "completions" && isPythonTruthy(payload["prompt"])) {
    return [messageText(payload["prompt"])];
  }
  return [];
}

function responsesInputItems(payload: Readonly<Record<string, unknown>>): unknown[] {
  const input = payload["input"];
  if (Array.isArray(input)) {
    return input;
  }
  return input === null || input === undefined ? [] : [input];
}

function responsesInputItemSummary(item: unknown): unknown {
  if (typeof item === "string" || !isRecord(item)) {
    return item;
  }
  const summary: Record<string, unknown> = {};
  for (const key of ["type", "role", "call_id", "name"] as const) {
    if (item[key] !== null && item[key] !== undefined) {
      summary[key] = item[key];
    }
  }
  const content = item["content"];
  if (Array.isArray(content)) {
    summary["content"] = (content as unknown[]).map((contentItem) => {
      if (!isRecord(contentItem)) {
        return contentItem;
      }
      return Object.fromEntries(
        ["type", "text", "arguments", "call_id"]
          .filter((key) => contentItem[key] !== null && contentItem[key] !== undefined)
          .map((key) => [key, contentItem[key]]),
      );
    });
  } else if (content !== null && content !== undefined) {
    summary["content"] = messageText(content);
  }
  for (const key of ["output", "arguments"] as const) {
    if (item[key] !== null && item[key] !== undefined) {
      summary[key] = messageText(item[key]);
    }
  }
  return Object.keys(summary).length > 0 ? summary : messageText(item);
}

function responsesInputPrefix(payload: Readonly<Record<string, unknown>>): unknown[] {
  return responsesInputItems(payload)
    .filter((item) => !isTaskContextMessage(item))
    .slice(0, 6)
    .map((item) => responsesInputItemSummary(item));
}

function responsesFirstUserMessage(payload: Readonly<Record<string, unknown>>): unknown {
  for (const item of responsesInputItems(payload)) {
    if (!isRecord(item) || item["role"] !== "user" || isTaskContextMessage(item)) {
      continue;
    }
    const content = item["content"];
    if (Array.isArray(content)) {
      const texts = (content as unknown[])
        .filter(
          (contentItem) =>
            isRecord(contentItem) &&
            typeof contentItem["text"] === "string" &&
            contentItem["text"] !== "",
        )
        .map((contentItem) => (contentItem as Record<string, unknown>)["text"]);
      if (texts.length > 0) {
        return texts;
      }
    } else if (isPythonTruthy(content)) {
      return messageText(content);
    }
  }
  return undefined;
}

function claudeSystemMessages(payload: Readonly<Record<string, unknown>>): unknown[] {
  const system = payload["system"];
  return isPythonTruthy(system) ? [{ role: "system", content: claudeTaskValue(system) }] : [];
}

function claudeMessageSummary(message: Readonly<Record<string, unknown>>): Record<string, unknown> {
  const summary: Record<string, unknown> = {
    role: message["role"] ?? null,
    content: claudeTaskValue(message["content"]),
  };
  for (const key of ["name", "tool_use_id"] as const) {
    if (message[key] !== null && message[key] !== undefined) {
      summary[key] = message[key];
    }
  }
  return summary;
}

function claudeMessages(payload: Readonly<Record<string, unknown>>): unknown[] {
  const messages = payload["messages"];
  if (!Array.isArray(messages)) {
    return [];
  }
  const compacted: unknown[] = [];
  for (const message of messages as unknown[]) {
    if (isRecord(message) && !isTaskContextMessage(message)) {
      compacted.push(claudeMessageSummary(message));
    }
  }
  return compacted;
}

function claudeFirstUserMessage(payload: Readonly<Record<string, unknown>>): unknown {
  const messages = payload["messages"];
  if (!Array.isArray(messages)) {
    return undefined;
  }
  for (const message of messages as unknown[]) {
    if (isRecord(message) && message["role"] === "user" && !isTaskContextMessage(message)) {
      return claudeTaskValue(message["content"]);
    }
  }
  return undefined;
}

function claudeTaskValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => claudeTaskValue(item));
  }
  if (isRecord(value)) {
    return Object.fromEntries(
      Object.keys(value)
        .filter((key) => !CLAUDE_TRANSIENT_TASK_FIELDS.has(key))
        .sort()
        .map((key) => [key, claudeTaskValue(value[key])]),
    );
  }
  return value === undefined ? null : value;
}

function messageText(value: unknown): unknown {
  if (value === undefined) {
    return null;
  }
  if (Array.isArray(value)) {
    return value.map((item) => messageText(item));
  }
  if (isRecord(value)) {
    return Object.fromEntries(
      Object.keys(value)
        .sort()
        .map((key) => [key, messageText(value[key])]),
    );
  }
  return value;
}

function chatSystemMessages(payload: Readonly<Record<string, unknown>>): unknown[] {
  const messages = payload["messages"];
  if (!Array.isArray(messages)) {
    return [];
  }
  const systemMessages: unknown[] = [];
  for (const message of messages as unknown[]) {
    if (isRecord(message) && (message["role"] === "system" || message["role"] === "developer")) {
      systemMessages.push({
        role: message["role"] ?? null,
        content: messageText(message["content"]),
      });
    }
  }
  return systemMessages;
}

function chatMessages(payload: Readonly<Record<string, unknown>>, limit?: number): unknown[] {
  const messages = payload["messages"];
  if (!Array.isArray(messages)) {
    return [];
  }
  const compacted: unknown[] = [];
  for (const message of messages) {
    if (!isRecord(message) || isTaskContextMessage(message)) {
      continue;
    }
    compacted.push({
      role: message["role"] ?? null,
      content: messageText(message["content"]),
      name: message["name"] ?? null,
      tool_call_id: message["tool_call_id"] ?? null,
    });
    if (limit !== undefined && compacted.length >= limit) {
      break;
    }
  }
  return compacted;
}

function chatFirstUserMessage(payload: Readonly<Record<string, unknown>>): unknown {
  const messages = payload["messages"];
  if (!Array.isArray(messages)) {
    return undefined;
  }
  for (const message of messages) {
    if (isRecord(message) && message["role"] === "user" && !isTaskContextMessage(message)) {
      return messageText(message["content"]);
    }
  }
  return undefined;
}

export function isTaskContextMessage(item: unknown): boolean {
  if (!isRecord(item)) {
    return false;
  }
  const text = contentText(item["content"]).trimStart();
  return [
    "<environment_context>",
    "<permissions instructions>",
    "<app-context>",
    "# Codex desktop context",
  ].some((prefix) => text.startsWith(prefix));
}

function contentText(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }
  if (Array.isArray(value)) {
    return value
      .flatMap((item) => {
        if (typeof item === "string") {
          return [item];
        }
        return isRecord(item) && typeof item["text"] === "string" ? [item["text"]] : [];
      })
      .join("\n");
  }
  return isRecord(value) && typeof value["text"] === "string" ? value["text"] : "";
}

function responseUsage(payload: unknown): unknown {
  if (!isRecord(payload)) {
    return undefined;
  }
  const streamSummary = payload["stream_summary"];
  if (isRecord(streamSummary) && isPythonTruthy(streamSummary["usage"])) {
    return streamSummary["usage"];
  }
  if (isPythonTruthy(payload["usage"])) {
    return payload["usage"];
  }
  const response = payload["response"];
  return isRecord(response) && isPythonTruthy(response["usage"]) ? response["usage"] : undefined;
}

function tokenValue(usage: Readonly<Record<string, unknown>>, key: string): number | undefined {
  const value = usage[key];
  return typeof value === "number" && Number.isInteger(value) ? value : undefined;
}

function isPythonTruthy(value: unknown): boolean {
  if (value === null || value === undefined || value === false || value === 0 || value === "") {
    return false;
  }
  if (Array.isArray(value)) {
    return value.length > 0;
  }
  if (isRecord(value)) {
    return Object.keys(value).length > 0;
  }
  return true;
}
